redgenes/bakta_annotations.py

```diff
@@ -0,0 +1,189 @@
+import pandas as pd
+from pathlib import Path
+from collections import defaultdict
+from sql_initialize_db import initialize_db
+from utils import (
+    copy_and_unzip,
+    run_command_and_check_outputs,
+    read_gff_file,
+    extract_gff_info,
+    process_gff_info,
+)
+from exceptions import (
+    InvalidInputTsv,
+    BaktaError,
+)
+from sql_connection import TRN
+
+bakta_version = 1
+bakta_commands = 1
+
+# Ensure all necessary imports are included
+def process_dbxref(dbxref):
+    """
+    Process dbxref entry and return a structured dictionary.
+    """
+    entry_dict = defaultdict(list)
+    if pd.notna(dbxref):
+        entries = dbxref.split(';')
+        for entry in entries:
+            if ':' in entry:
+                entry_type, entry_value = entry.split(':', 1)
+                entry_dict[entry_type.strip()].append(entry_value.strip())
+    return entry_dict
+
+def insert_dbxref_info(bakta_accesion, dbxref_data):
+    """
+    Insert dbxref information into corresponding tables.
+    """
+    for dbxref_type, accession_list in dbxref_data.items():
+        table_name = dbxref_type.lower()  # Assuming table names are lowercase versions of dbxref types
+        if table_name in ['kegg', 'refseq', 'uniparc', 'uniref', 'so', 'pfam']:  # Check if table_name is valid
+            sql = f"INSERT INTO {table_name} (bakta_accession, {table_name.upper()}) VALUES (?, ?)"
+            for accession in accession_list:
+                TRN.add(sql, [bakta_accesion, accession])
+    TRN.execute()
+
+def run_bakta(input_fasta: Path, outdir: Path, bakta_db):
+    """
+    Run Bakta and return the path to the output GFF and the command string.
+    """
+    stem = input_fasta.stem
+    bakta_output_tsv = outdir / f"{stem}.tsv"
+    commands = [
+        "bakta",
+        "--skip-plot",
+        "--keep-contig-headers",
+        "--db", str(bakta_db),
+        "--output", str(outdir),
+        "--prefix", str(stem),
+        "--locus-tag", str(stem), 
+        "--force",
+        str(input_fasta)
+    ]
+    _, commands_str = run_command_and_check_outputs(
+        commands, BaktaError, [bakta_output_tsv]
+    )
+    return bakta_output_tsv, commands_str
+
+def extract_md_info(tsv_file: Path):
+    """
+    Load data from TSV and check data integrity.
+    """
+    if not tsv_file.exists():
+        raise FileNotFoundError(f"TSV file not found: {tsv_file}")
+    df = pd.read_csv(tsv_file, sep="\t", dtype=str)
+    if df.shape[1] != 40:
+        raise InvalidInputTsv(f"Invalid input TSV: {tsv_file}")
+    df["annotation_date"] = pd.to_datetime(df["annotation_date"], errors="coerce")
+    dtype_map = {
+        "taxid": "Int64",
+        "species_taxid": "Int64",
+        "gc_percent": float,
+        "replicon_count": "Int64",
+        "scaffold_count": "Int64",
+        "contig_count": "Int64",
+        "total_gene_count": "Int64",
+        "protein_coding_gene_count": "Int64",
+        "non_coding_gene_count": "Int64",
```

Comment:
I think the ints can all be unsigned as, for example, a negative tax ID doesn't make sense. Is 64 bit needed or is 32 bit sufficient for most / many of these?